service/graph.go

```diff
@@ -31,21 +31,209 @@ var _ pipelines = (*pipelinesGraph)(nil)
+		}
+	}
+	pipelines.createNodes(set)
+	pipelines.createEdges()
+	return pipelines, pipelines.buildNodes(ctx, set)
+}
+
+// Creates a node for each instance of a component and adds it to the graph
+func (g *pipelinesGraph) createNodes(set pipelinesSettings) {
+	// Keep track of connectors and where they are used. (map[connectorID][]pipelineID)
+	connectorsAsExporter := make(map[component.ID][]component.ID)
+	connectorsAsReceiver := make(map[component.ID][]component.ID)
+
+	for pipelineID, pipelineCfg := range set.PipelineConfigs {
+		pipe := g.pipelines[pipelineID]
+		for _, recvID := range pipelineCfg.Receivers {
+			if set.Connectors.IsConfigured(recvID) {
+				connectorsAsReceiver[recvID] = append(connectorsAsReceiver[recvID], pipelineID)
+				continue
+			}
+			pipe.addReceiver(g.createReceiver(pipelineID, recvID))
+		}
+		for _, procID := range pipelineCfg.Processors {
+			pipe.addProcessor(g.createProcessor(pipelineID, procID))
+		}
+		for _, exprID := range pipelineCfg.Exporters {
+			if set.Connectors.IsConfigured(exprID) {
+				connectorsAsExporter[exprID] = append(connectorsAsExporter[exprID], pipelineID)
+				continue
+			}
+			pipe.addExporter(g.createExporter(pipelineID, exprID))
+		}
+	}
+
+	for connID, exprPipelineIDs := range connectorsAsExporter {
+		for _, eID := range exprPipelineIDs {
+			for _, rID := range connectorsAsReceiver[connID] {
+				connNode := g.createConnector(eID, rID, connID)
+				g.pipelines[eID].addExporter(connNode)
+				g.pipelines[rID].addReceiver(connNode)
+			}
+		}
+	}
+}
+
+func (g *pipelinesGraph) createReceiver(pipelineID, recvID component.ID) *receiverNode {
+	rcvrNode := newReceiverNode(pipelineID, recvID)
+	if node := g.componentGraph.Node(rcvrNode.ID()); node != nil {
+		return node.(*receiverNode)
+	}
+	g.componentGraph.AddNode(rcvrNode)
+	return rcvrNode
+}
+
+func (g *pipelinesGraph) createProcessor(pipelineID, procID component.ID) *processorNode {
+	procNode := newProcessorNode(pipelineID, procID)
+	g.componentGraph.AddNode(procNode)
+	return procNode
+}
+
+func (g *pipelinesGraph) createExporter(pipelineID, exprID component.ID) *exporterNode {
+	expNode := newExporterNode(pipelineID, exprID)
+	if node := g.componentGraph.Node(expNode.ID()); node != nil {
+		return node.(*exporterNode)
+	}
+	g.componentGraph.AddNode(expNode)
+	return expNode
+}
+
+func (g *pipelinesGraph) createConnector(exprPipelineID, rcvrPipelineID, connID component.ID) *connectorNode {
+	connNode := newConnectorNode(exprPipelineID.Type(), rcvrPipelineID.Type(), connID)
+	if node := g.componentGraph.Node(connNode.ID()); node != nil {
+		return node.(*connectorNode)
+	}
+	g.componentGraph.AddNode(connNode)
+	return connNode
+}
+
+func (g *pipelinesGraph) createEdges() {
+	for pipelineID, pg := range g.pipelines {
+		fanOutToExporters := newFanOutNode(pipelineID)
+		for _, exporter := range pg.exporters {
+			g.componentGraph.SetEdge(g.componentGraph.NewEdge(fanOutToExporters, exporter))
+		}
+
+		if len(pg.processors) == 0 {
```

Comment:
Not sure it's quite what you were looking for, but I changed it so that every pipeline has a capabilities node and fanout node. Now creating edges has no special cases.

> n the building make its base consumer == { the processor if only one processor; the fanOutToExporters if no processor; 

This is already the case because the edge from capabilities node points to the correct next consumer, either the first processor or the fanout node.

> create a real capability consumer if more than one processor and at least one mutates

You might be referring to something more specific here, but I think the capabilities node already _is_ a real capability consumer in the sense that it is a base consumer and also can describe its capabilities (aggregated from the processors).